crates/sparrow-runtime/src/prepare/column_behavior.rs

```diff
@@ -0,0 +1,525 @@
+use anyhow::Context;
+use arrow::{
+    array::ArrayRef,
+    datatypes::{DataType, Field, SchemaRef},
+    record_batch::RecordBatch,
+};
+use sparrow_core::context_code;
+
+use super::PrepareMetadata;
+use std::{borrow::BorrowMut, sync::Arc};
+
+use anyhow::anyhow;
+use arrow::array::{Array, UInt64Array};
+
+use error_stack::{IntoReport, IntoReportCompat, ResultExt};
+use sparrow_core::utils::make_null_array;
+use sparrow_kernels::order_preserving_cast_to_u64;
+
+use crate::{execute::key_hash_inverse::ThreadSafeKeyHashInverse, prepare::Error};
+
+/// Defines how each column in the resulting prepared batch
+/// is computed.
+#[derive(Debug)]
+pub enum ColumnBehavior {
+    /// Cast the given column to the given data type.
+    Cast {
+        index: usize,
+        data_type: DataType,
+        nullable: bool,
+    },
+    /// Perform an "order preserving" cast from a primitive number to u64.
+    OrderPreservingCastToU64 { index: usize, nullable: bool },
+    /// Reference the given column.
+    Reference { index: usize, nullable: bool },
+    /// Hash the given column.
+    ///
+    /// This field is used for preparing the key column during preparation.
+    PrepareEntityKey { index: usize, nullable: bool },
```

Comment:
In the comments, we should explain why these are different. Otherwise, we're just saying "we have different behavior during preparation and execution, go read the code to find out what that means". Ideally, we could even name these more specifically to how they differ.

Reply:
Added a few comments

Reply:
Why is that a different column behavior? If I remember correctly -- metadata needs the key and key hash columns to update the metadata. I think instead of having the column behavior add to the metadata, we should be able to do that *after* preparing. Specifically:

1. Use the non-metadata populating column and prepare the columns.
2. The result of this should be a batch that contains a `_key_hash` column (prepared using `ExecuteEntityKey`) and a `<key>` column (for whatever the name of the underlying key was).
3. Thus, the batch can be added to the metadata *after* being prepared, by extracting those two columns.

In addition to getting rid of the need to distinguish between these two cases, it moves the maintenance of the metadata outside of the "column behavior" which seems a clearer place to have it.

Reply:
Yeah, I noted that in the comment 😢 . I can make that change.